spec/models/manageiq/providers/embedded_terraform/automation_manager/configuration_script_source_spec.rb

```diff
@@ -0,0 +1,457 @@
+    #       :op_name => "#{described_class::FRIENDLY_NAME} #{action}",
+    #       :op_arg  => "(#{op_arg.except(:name).map { |k, v| "#{k}=#{v}" }.join(', ')})",
+    #       :tower   => "EMS(manager_id=#{manager.id})"
+    #     }
+    #   }
+    # end
+  end
+
+  describe "git_repository interaction" do
+    let(:auth) { FactoryBot.create(:embedded_terraform_scm_credential) }
+    let(:configuration_script_source) do
+      described_class.create!(
+        :name           => "foo",
+        :scm_url        => "https://example.com/foo.git",
+        :authentication => auth
+      )
+    end
+
+    it "on .create" do
+      configuration_script_source
+
+      git_repository = GitRepository.first
+      expect(git_repository.name).to eq "foo"
+      expect(git_repository.url).to eq "https://example.com/foo.git"
+      expect(git_repository.authentication).to eq auth
+
+      expect { configuration_script_source.git_repository }.to_not make_database_queries
+      expect(configuration_script_source.git_repository_id).to eq git_repository.id
+    end
+
+    it "on .new" do
+      configuration_script_source = described_class.new(
+        :name           => "foo",
+        :scm_url        => "https://example.com/foo.git",
+        :authentication => auth
+      )
+
+      expect(GitRepository.count).to eq 0
+
+      attached_git_repository = configuration_script_source.git_repository
+
+      git_repository = GitRepository.first
+      expect(git_repository).to eq attached_git_repository
+      expect(git_repository.name).to eq "foo"
+      expect(git_repository.url).to eq "https://example.com/foo.git"
+      expect(git_repository.authentication).to eq auth
+
+      expect { configuration_script_source.git_repository }.to_not make_database_queries
+      expect(configuration_script_source.git_repository_id).to eq git_repository.id
+    end
+
+    it "errors when scm_url is invalid" do
+      expect do
+        configuration_script_source.update!(:scm_url => "invalid url")
+      end.to raise_error(ActiveRecord::RecordInvalid)
+    end
+
+    it "syncs attributes down" do
+      configuration_script_source.name = "bar"
+      expect(configuration_script_source.git_repository.name).to eq "bar"
+
+      configuration_script_source.scm_url = "https://example.com/bar.git"
+      expect(configuration_script_source.git_repository.url).to eq "https://example.com/bar.git"
+
+      configuration_script_source.authentication = nil
+      expect(configuration_script_source.git_repository.authentication).to be_nil
+    end
+
+    it "persists attributes down" do
+      configuration_script_source.update!(:name => "bar")
+      expect(GitRepository.first.name).to eq "bar"
+
+      configuration_script_source.update!(:scm_url => "https://example.com/bar.git")
+      expect(GitRepository.first.url).to eq "https://example.com/bar.git"
+
+      configuration_script_source.update!(:authentication => nil)
+      expect(GitRepository.first.authentication).to be_nil
```

Comment:
Are any of the tests in the `git_repository interaction` describe/context specific to terraform?  

I only see display_name, sync, template_name_from_git_repo_url, and find_templates_in_git_repo.  display_name is a simple method to test.  The others need to be tested.  I think any other tests should be covered by the parent classes' tests.

I can help on the sync and find_templates_in_git_repo tests. I'll do that tomorrow. I provided the basics for template_name_from_git_repo_url.

Reply:
After you review/pull in my test changes, I think these should be removed if they're not testing the methods or side effects of the methods in the file.  Basically, if you can comment out the methods in the file, and these tests still pass, they're tests for the parent or grandparent class.